### PR TITLE
feat(tokens): script-aware estimation, single source of truth, provider registry (#525, #530, #493)

### LINKED ISSUE #493: Add provider-calibrated token estimation with a calibration benchmark and estimator-path observability
## Summary

Make token estimation accuracy measurable and provider-aware: add a calibration benchmark comparing the built-in estimators against real provider tokenizers, support per-provider estimator registration, and surface which estimator path produced each count.

## Why this matters

Every budget decision in the library — phase budgets, firewall thresholds, card token caps, savings claims — flows through token counts. The default counter is `tiktoken` (OpenAI-family vocabularies), while many deployments target Anthropic or Gemini models whose tokenization differs; the offline fallback is `len(text)//4`, which can be off by 20–50%. Quantifying the divergence, documenting it, and letting integrators register accurate counters per provider makes budget enforcement and published token-savings numbers trustworthy across the whole provider matrix the adapters already support.

## Current evidence

- `src/contextweaver/tokens.py`: `TiktokenEstimator` default with `@lru_cache(maxsize=16)`; fallback `CharDivFourEstimator` (`src/contextweaver/protocols.py` lines ~51–55) returns `len(text)//4`.
- `src/contextweaver/routing/cards.py` (~line 108): independent `len(text)//4` fallback inside card budgeting.
- Nothing records which estimator path produced a count — `BuildStats`/`FirewallStats` report numbers without provenance.
- Open issue #413 covers *documenting* the estimator plug point; this issue covers calibration, registry, and observability.

## Proposed implementation

1. **Calibration benchmark** (`benchmarks/token_calibration.py`): a fixed multilingual/multi-shape text corpus (prose, JSON, code, logs) counted by (a) `CharDivFourEstimator`, (b) tiktoken encodings, and — opt-in, credentialed, never CI — (c) provider count-tokens APIs (Anthropic/Gemini). Output a divergence table (mean/max relative error per corpus shape) committed to the docs.
2. **Estimator registry**: `register_estimator(name, estimator)` + `get_token_counter(model_or_provider)` resolution in `tokens.py`, so an Anthropic-accurate counter (user-supplied or heuristic-calibrated) can be selected by configuration without code changes. Keep `tiktoken` the default.
3. **Calibrated fallback**: replace the bare `//4` heuristic with a shape-aware heuristic (distinct ratios for code/JSON vs prose) derived from the calibration corpus; keep it deterministic and dependency-free.
4. **Observability**: record the estimator identifier in `BuildStats` (e.g. `token_estimator: "tiktoken/cl100k_base"` vs `"heuristic/v2"`), so budget overshoots can be attributed. Unify the `cards.py` fallback through `tokens.py`.

## Example prompt, schema, or interface

```python
from contextweaver.tokens import register_estimator, get_token_counter

register_estimator("anthropic", my_anthropic_counter)   # wraps client.count_tokens
counter = get_token_counter("anthropic")
# BuildStats.token_estimator == "anthropic"
```

## Acceptance criteria

- A committed divergence table documents heuristic and tiktoken error versus at least two provider tokenizers across ≥4 corpus shapes.
- Estimators are registrable and selectable by provider/model name; default behavior unchanged.
- `BuildStats` reports the estimator path used.
- The card-rendering fallback delegates to `tokens.py` (single source of truth).

## Test and evaluation plan

- Unit tests for registry resolution, fallback ordering, and stats stamping.
- Calibration benchmark runs offline-deterministic for (a)/(b); provider-API leg is opt-in via env var and skips cleanly (mirrors `CW_E2E_LLM` pattern).
- Regression: existing budget tests and committed scorecards unchanged under the default estimator.

## Migration notes

Replacing the raw `//4` fallback changes counts in tiktoken-less environments; document in CHANGELOG and regenerate any affected committed artifacts. Estimator-name field on `BuildStats` is additive.

## Risks and tradeoffs

Provider tokenizers change over time — the divergence table needs a recorded date/model version. A shape-aware heuristic is still a heuristic; the table must present it honestly. Registry indirection adds minor complexity to a deliberately small module.

## Suggested labels

performance, provider-abstraction, reliability, area/context

### LINKED ISSUE #525: Add non-ASCII regression coverage and fix CJK token under-estimation in the heuristic fallback
## Summary

Add regression coverage for non-ASCII content (CJK, emoji, RTL, combining characters) across tokenization, budgeting, dedup, card rendering, and the CLI — and make the offline heuristic token estimator script-aware so CJK text is not under-counted by ~4×.

Priority: P0 · Confidence: High

## Why this matters

Budget enforcement is the product's core guarantee. The offline/air-gapped fallback estimates tokens as `len(text) // 4`, which is roughly right for English but wildly wrong for CJK scripts where one character is typically ≈1 token or more — so a Japanese or Chinese transcript can overrun a phase budget by ~4× exactly when the exact tokenizer is unavailable. Separately, the test suite has essentially no non-ASCII coverage, leaving unicode behavior in dedup, tag capping (`t[:24]` can split graphemes), and Rich CLI rendering unverified for a global user base.

## Current evidence

- `src/contextweaver/tokens.py` (~lines 67–79): character-heuristic fallback divides length by 4 with no script awareness; `src/contextweaver/context/sensitivity.py:84` hard-codes the same `// 4` for redaction placeholders.
- Test audit: only one fallback-related test found (`tests/test_protocols.py::test_tiktoken_fallback_when_encoding_unavailable`); zero tests with CJK, emoji, or RTL content across `tests/` (verified by search).
- `src/contextweaver/routing/cards.py:176`: `t[:24]` tag truncation operates on code points — multi-code-point graphemes (emoji ZWJ sequences) can be split.
- `docs/troubleshooting.md` documents the offline-tiktoken warning and calls the fallback "deterministic" — true, but its accuracy caveat is undocumented.

## External context

Common token-estimation guidance: ~4 chars/token holds for English; CJK text typically runs ~1+ token per character under cl100k-family encodings. A simple Unicode-range-based correction (counting CJK/Hangul/Kana code points at ~1 token each) is a standard cheap improvement and stays fully deterministic.

## Proposed implementation

1. Make the heuristic estimator script-aware in `tokens.py`: count code points in CJK/Kana/Hangul ranges at a per-char rate (≈1.0, calibrate against tiktoken on sample corpora) and remaining text at the existing `len/4` rate. Keep it deterministic and dependency-free.
2. Replace the hard-coded `// 4` in `sensitivity.py` with the configured estimator (small, related fix — single source of truth per `tokens.py`'s charter).
3. Add a non-ASCII regression suite: budgeting with CJK transcripts (offline-fallback mode), dedup/Jaccard on unicode text, card rendering with emoji tags and RTL descriptions, CLI `route`/`build` smoke with unicode catalogs, serde round-trips.
4. Document fallback accuracy expectations in `docs/troubleshooting.md`.

## AI-agent execution notes

- Inspect first: `tokens.py` (estimator + counter wiring), `_utils.py` `tokenize()` (unicode behavior), `context/sensitivity.py:83-84`, `routing/cards.py` truncations, `tests/test_tokens.py`, `tests/test_protocols.py`.
- Run: `pytest tests/test_tokens.py tests/test_dedup.py tests/test_cards.py -q`; calibrate by comparing against `tiktoken.cl100k_base` locally; `make ci`.
- Edge cases: mixed-script text; surrogate pairs/emoji; empty strings; very long CJK strings (no perf cliff — single pass).
- Preserve: determinism; existing English-text estimates should not change dramatically (state the acceptable delta in tests); tiktoken path untouched.
- Do not add any new dependency for script detection — use `unicodedata`/range checks.

## Acceptance criteria

- For a representative CJK corpus, the fallback estimate is within ±30% of tiktoken's count (vs ~-75% today), asserted by a committed-fixture test.
- Redaction placeholders use the configured estimator.
- New unicode regression tests pass across tokenization, dedup, cards, serde, and CLI smoke paths.

## Test plan

Fixture corpora (EN/JA/ZH/KO/AR/emoji) with expected ranges; offline-mode test forcing the fallback; golden-prompt check confirming English-path stability; full `make ci`.

## Documentation plan

Accuracy note in `docs/troubleshooting.md` (offline tiktoken section) and `tokens.py` docstrings; CHANGELOG (behavior change in fallback estimates).

## Migration and compatibility notes

Fallback token estimates change for non-Latin text (more accurate, generally higher), which can change selection outcomes in offline mode — that is the fix. CHANGELOG note required; exact-tokenizer (default online) behavior is unchanged.

## Risks and tradeoffs

Any heuristic remains approximate (document the bands); slightly more CPU per estimate (single linear scan — negligible). Alternative of bundling a tokenizer for offline use was rejected: dependency cost contradicts the minimal-core policy.

## Suggested labels

testing, reliability, area/context, performance

### LINKED ISSUE #530: Use the configured TokenEstimator for redaction placeholders and other hard-coded estimates
## Summary

Route the remaining hard-coded `len(text) // 4` token estimates — notably the sensitivity-redaction placeholder — through the configured token counter, so every number that feeds budget enforcement comes from one source of truth.

Priority: P2 · Confidence: High

## Why this matters

`tokens.py` exists precisely so "firewall / FirewallStats numbers are computed through the same counter" (AGENTS.md module map, issue #405 lineage). Stray `// 4` literals bypass that contract: estimates disagree with the canonical counter, and any future estimator improvement (e.g., script-aware fallback) silently doesn't apply to these paths. Small fix, meaningful consistency win in security-adjacent code.

## Current evidence

- `src/contextweaver/context/sensitivity.py:83-84`:
  `placeholder = f"[REDACTED: {item.sensitivity.value}]"` followed by `token_estimate=len(placeholder) // 4` — a manual heuristic in a security-grade module, bypassing `contextweaver.tokens`.
- `src/contextweaver/tokens.py` provides `count()` / `get_token_counter()` as the canonical path.
- A repo-wide grep for `// 4` should be part of this change to catch any sibling literals (the audit flagged sensitivity.py specifically; others may exist).

## External context

Not required for this issue.

## Proposed implementation

1. Inventory all hard-coded character-division estimates under `src/contextweaver/` (`grep -rn "// 4"` and similar).
2. In `sensitivity.py`, compute the placeholder's `token_estimate` via the configured estimator. Design decision: the redaction function may need access to the estimator — thread it through the existing call path from the pipeline (the sensitivity stage runs inside the build, where the manager's estimator is available) rather than importing a global; follow `.claude/rules/sensitivity.md` caution (no behavioral weakening — only the estimate's source changes).
3. Repeat for any other call sites found; leave `tokens.py`'s own fallback heuristic as the single allowed implementation.

## AI-agent execution notes

- Inspect first: `context/sensitivity.py` (full module — security-grade, read `.claude/rules/sensitivity.md` and `docs/agent-context/invariants.md` first), `tokens.py`, `context/build.py` (how the estimator reaches stages).
- Run: `pytest tests/test_sensitivity.py tests/test_sensitivity_fixtures.py tests/test_tokens.py -q`; `make ci`.
- Add tests: redacted item's `token_estimate` equals the configured counter's output for the placeholder; a custom estimator passed to the manager is honored on redaction paths.
- Preserve: redaction text/format unchanged; drop/redact decision logic untouched; determinism.
- Do not refactor sensitivity.py structure (explicitly protected).

## Acceptance criteria

- No `// 4`-style literal estimates remain outside `tokens.py`.
- Redaction placeholder estimates come from the configured estimator, proven by a custom-estimator test.
- Sensitivity behavior is otherwise byte-identical (existing fixture tests pass unchanged).

## Test plan

Unit tests above; the sensitivity fixture suite (`tests/fixtures/sensitivity/*`) green; golden prompts unchanged (placeholder text identical; only the internal estimate differs).

## Documentation plan

CHANGELOG note; touch `tokens.py` docstring if its "single source of truth" charter wording should name the sensitivity path.

## Migration and compatibility notes

Not expected to require migration — internal estimate sourcing; selection outcomes could shift by ±1 token in pathological budget-edge cases (CHANGELOG-worthy, nothing more).

## Risks and tradeoffs

Threading the estimator into the sensitivity stage touches security-grade code — keep the diff minimal and mechanically reviewable. The alternative (leave as-is) perpetuates two estimate sources, which gets worse once the fallback becomes script-aware.

## Suggested labels

area/context, reliability, good-first-ai-issue

### PR DESCRIPTION
**Pull request by dgenio**
## Summary

Unify token estimation behind `contextweaver.tokens` and fix the offline CJK
budget under-count. One combined PR for the token-estimation cluster:

Closes #525
Closes #530
Closes #493

## Changes

**#525 — script-aware heuristic + non-ASCII coverage**
- New `HeuristicEstimator` (`protocols.py`): counts CJK / Kana / Hangul / emoji
  code points at ≈1 token/char and the rest at `len // 4`. Deterministic,
  stdlib-only (integer range checks, single linear scan). **Identical to
  `len // 4` for Latin/ASCII**, so English estimates do not change; fixes the
  ~4× offline under-count on dense scripts. Now the default `ContextManager`
  estimator and the `TiktokenEstimator` offline fallback.
- New `tests/test_unicode_regression.py`: CJK/emoji/RTL across tokenization,
  budgeting, dedup, card rendering, serde, and an in-process build.

**#530 — one source of truth for token counts**
- Removed the stray `len // 4` literals (sensitivity redaction placeholder,
  firewall summary item, `routing/cards` + `routing/packer` fallbacks,
  memory-source costing) — they now route through the configured estimator /
  `tokens.py`. `CharDivFourEstimator` remains the lone `len // 4` primitive.
- The sensitivity stage receives the manager's estimator, so a **custom counter
  is honoured on redaction paths** (proven by a test). ASCII placeholder
  estimates are unchanged (`heuristic == len // 4` for ASCII), so existing
  sensitivity fixtures pass unchanged.

**#493 — provider-calibrated estimation + observability**
- `tokens.register_estimator()` / `registered_estimators()` / `estimator_name()`
  + provider/model-name resolution in `get_token_counter()`; `tiktoken` stays
  the default.
- New additive `BuildStats.token_estimator` field records the estimator path
  (`"tiktoken/cl100k_base"`, `"heuristic/v2"`, or a registered provider name);
  schemas regenerated.
- `routing/cards.count_tokens` now delegates to `tokens.py` (deletes the
  duplicate encoder + fallback).
- `benchmarks/token_calibration.py` + `make token-calibration` render
  `docs/token_calibration.md` across ≥4 corpus shapes; provider `count_tokens`
  legs are opt-in via `CW_TOKEN_CALIBRATION_PROVIDERS` and never run in CI.
- Docs: `AGENTS.md` (module map for `tokens.py` / `protocols.py` / `BuildStats`),
  `CHANGELOG.md`, `docs/troubleshooting.md` (script-aware default + provider
  selection; also fixed a stale `token_estimator=` kwarg → `estimator=`),
  mkdocs nav.

## Checklist

- [x] Tests added or updated for every new/changed public function
- [x] `make ci` passes locally (fmt + lint + type + test + schemas-check + example + demo) — see "How verified" for the one offline-only caveat
- [x] `CHANGELOG.md` updated under `## [Unreleased]`
- [x] Docstrings added for all new public APIs (Google-style)
- [ ] Every modified module stays ≤ 300 lines — `protocols.py` (664) and `cards.py` (408) exceed the guideline, but **both were already over before this PR** (this change actually *removes* ~40 lines from `cards.py`). The new estimators belong with the other protocol implementations per the module map; decomposing these pre-existing large modules is out of scope here.
- [x] Related issue linked in the summary above
- [x] Agent-facing docs updated (`AGENTS.md`, `docs/troubleshooting.md`)

## How verified

- `ruff format --check` + `ruff check src/ tests/ examples/ scripts/ benchmarks/token_calibration.py` — clean.
- `mypy src/` — Success: no issues in 126 source files.
- `python scripts/gen_schemas.py --check` — schemas up to date (after regen).
- `pytest` — **2021 passed, 30 skipped, 1 xfailed, 1 failed**. The single
  failure is `tests/test_mcp_serve_cli.py::test_serve_dry_run_writes_catalog_diagnostic_event`,
  which is a **pre-existing offline-only artifact** (verified: it fails
  identically on `main` in this sandbox). It asserts empty stderr, but
  `tiktoken` emits its offline-fallback warning because the sandbox cannot
  download the BPE encoding (HTTP 403). It passes in CI where the tiktoken
  cache is warm.
- `make demo` + `examples/{minimal_loop,before_after,routing_demo,full_agent_loop}.py` — pass.
- `mkdocs build --clean` — clean.

## Notes for reviewers

- **Offline calibration caveat (action needed in a connected env):** this PR was
  developed in a sandbox where `tiktoken` cannot fetch `cl100k_base` (HTTP 403),
  so the committed `docs/token_calibration.md` has real heuristic + char columns
  but its `tiktoken`/provider-error columns read `n/a`. Re-run
  `make token-calibration` with a warmed `TIKTOKEN_CACHE_DIR` to populate those
  columns. No tokenizer numbers were invented.
- The default estimator changed from `CharDivFourEstimator` to
  `HeuristicEstimator`. Behaviour is byte-identical for Latin/ASCII; offline
  non-Latin estimates become more accurate (and generally higher), which can
  shift selection outcomes in offline mode by design (called out in CHANGELOG).
- `sensitivity.py` is security-grade: the diff is mechanical (only the
  placeholder's estimate *source* changes; drop/redact decisions, redaction
  text, and defaults are untouched).

## Reproducibility (scoring / context-pipeline changes)

Token counts feed budget/selection, so this can move `prompt_tokens` on
non-Latin content offline (the intended fix). For Latin/ASCII corpora — which
the committed scorecard/golden fixtures use — estimates are unchanged; the
golden route-prompt fixture only needed the additive `token_estimator` key.
Run `make benchmark-matrix && make scorecard` in a tiktoken-warm environment to
refresh the sticky benchmark-delta comment.

---
_Generated by [Claude Code](https://claude.ai/code/session_01PnouojFtFEraDod2zN2sh5)_

### CODE REVIEW

**Comment by dgenio** [comment]
Thanks — all three addressed in 8fa0005:

- **`tokens.py` (memoisation docs):** reworded `get_token_counter` / `_tiktoken_counter` to state the cache key is the raw `model` string, so two names sharing an encoding are cached separately. (No behaviour change.)
- **`memory_source.py` (per-entry allocation):** `_estimate_cost` now reuses a module-level `_DEFAULT_COUNTER` instead of constructing a `HeuristicEstimator` per entry.
- **`token_calibration.py` (registry corrupting the baseline):** the tiktoken reference column now instantiates `TiktokenEstimator(_TIKTOKEN_MODEL)` directly, bypassing the registry, so a counter registered under the encoding name can't replace the baseline.

Re-ran `ruff`, `mypy`, and the affected suites (token/memory/calibration/protocols/unicode) — green; the regenerated calibration doc is unchanged offline.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/contextweaver/tokens.py [hunk 1]
```diff
@@ -46,22 +60,75 @@
...
     Returns:
-        A :class:`~contextweaver.protocols.TokenEstimator`.  The result is
-        memoised per *model* so repeated calls reuse the loaded encoding.  In
-        offline environments the underlying :class:`TiktokenEstimator` falls
-        back to the character heuristic; callers always get a usable counter.
+        A :class:`~contextweaver.protocols.TokenEstimator`.  Registered names
+        resolve to their registered counter; otherwise the result is a
+        tiktoken-backed counter, memoised per encoding so repeated calls reuse
+        the loaded encoding.  In offline environments the underlying
+        :class:`TiktokenEstimator` falls back to the script-aware heuristic;
```
[the remark this answers is not in the document]
Comment (dgenio) [reply src/contextweaver/tokens.py]: Good catch — the docstring was misleading. Reworded in `8fa0005`: `_tiktoken_counter` and `get_token_counter` now state the cache key is the exact `model` string, so two model names sharing an encoding are cached as separate instances (`tokens.py:107-137`). Docs-only clarification — no behaviour change.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/contextweaver/context/memory_source.py [hunk 2]
```diff
@@ -14,15 +14,21 @@
...
-    estimate = estimator.estimate(text) if estimator is not None else len(text) // 4
-    return max(1, int(estimate))
+    """Return a positive token estimate for *text*.
+
+    Falls back to the canonical script-aware heuristic counter (issue #530)
+    rather than an inline ``len // 4`` literal when no estimator is supplied,
+    so every budget number flows through one source of truth.
+    """
+    counter = estimator if estimator is not None else heuristic_counter()
+    return max(1, int(counter.estimate(text)))
```
[the remark this answers is not in the document]
Comment (dgenio) [reply src/contextweaver/context/memory_source.py:35]: Fixed in `8fa0005` — `_estimate_cost` now reuses a module-level `_DEFAULT_COUNTER = heuristic_counter()` (`memory_source.py:24,34`) instead of constructing a `HeuristicEstimator` per call, so large memory sources no longer allocate one estimator per entry. Callers can still thread their own estimator in.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: benchmarks/token_calibration.py [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+    heuristic = HeuristicEstimator()
+    tiktoken_ok = _tiktoken_available()
+    providers = _provider_counters()
+
+    shapes: list[dict[str, object]] = []
+    for shape, samples in _CORPUS.items():
+        chars = sum(len(s) for s in samples)
+        h_tokens = sum(heuristic.estimate(s) for s in samples)
+        tt_tokens = (
+            sum(tokens.count(s, model=_TIKTOKEN_MODEL) for s in samples) if tiktoken_ok else None
```
[the remark this answers is not in the document]
Comment (dgenio) [reply benchmarks/token_calibration.py]: Agreed — routing the reference column through the registry would let a counter registered under the encoding name shadow the tiktoken baseline. Fixed in `8fa0005`: the benchmark now instantiates `TiktokenEstimator(_TIKTOKEN_MODEL)` directly and bypasses `get_token_counter` (`token_calibration.py:107-110`), so the baseline can't be corrupted.

---
_Generated by [Claude Code](https://claude.ai/code)_
